lib/deployment-bucket-endpoint.js

```diff
@@ -0,0 +1,14 @@
+'use-strict';
+
+const { endpoint } = require('aws-info');
+
+module.exports = function setDeploymentBucketEndpoint() {
+  const bucket = this.serverless.service.provider.deploymentBucket;
+  if ( bucket !== undefined ) {
+    return this.provider.request('S3', 'getBucketLocation', { Bucket: bucket }).then((response) => {
+       this.deploymentBucketEndpoint = endpoint('S3', response.LocationConstraint);
```

Comment:
The docs https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#getBucketLocation-property

Suggest that this may return an empty value or "EU" in some cases. I think we need some custom logic here to map those, because the aws-info library is expecting them to be always in region code (xx-yyyy-n)

I believe the empty case is:
> When the bucket's region is US East (N. Virginia), Amazon S3 returns an empty string for the bucket's region:

Not sure what "EU" is ;)

Reply:
Apparently you can create a bucket with that constraint:

`aws s3api --region eu-west-1 create-bucket --bucket testbucket.example.org --create-bucket-configuration LocationConstraint=EU`

which makes everything so very complicated ...

Reply:
@dougmoscrop so that leaves us with option 1) my initial PR? :)

Reply:
Ok after digging around a little bit I found a way you can find the region of a bucket. You can simply `HEAD` request any s3 endpoint for the bucket and you will be returned a `x-amz-bucket-region: <region>` header. This you can use, and even for `LocationConstraints=EU` it will return the real region! E.g. in the above example:

```sh
$ aws s3api --region eu-west-1 create-bucket --bucket testbucket.example.org --create-bucket-configuration LocationConstraint=EU
{
    "LocationConstraint": "EU"
}

$ curl http://s3.ap-northeast-1.amazonaws.com/testbucket.example.org -I
HTTP/1.1 301 Moved Permanently
x-amz-bucket-region: eu-west-1
x-amz-request-id: 2A4F03A65C9D55FE
x-amz-id-2: u/IYMEiWFsbKMupXGL7D8cvhfAbO+tMjo33HrizFojkKSNSOa1AU1eVPj3Vt7GMupDDtM8EFtFs=
Content-Type: application/xml
Transfer-Encoding: chunked
Date: Wed, 01 May 2019 08:39:19 GMT
Server: AmazonS3
```

The technology is there 🎉

Reply:
yep this looks great to me!

Reply:
Any preference what client to use for the HTTP Head request?

Reply:
Oh, you can't use provider.request('S3', 'headBucket') ?

Reply:
Does the SDK swallow the info or error?

Reply:
No I don't believe you can. The command has no output (tested with the CLI). Apart from that, the `headBucket` request will follow the redirect and send 2 requests if the regions differ, we only need one to find out the region though. So HTTP Head is the best option imho

Reply:
I hear you but won't the raw http side need sigv4 and other complexity to get creds and so on? If I gives the info we need the extra call us probably worth the trade off

Reply:
Nevermind - you just used curl

What if we just assume the bucket is in the same region as the deployment and support a custom.bootstrap.bucketRegion setting?

Reply:
What's `custom.bootstrap.bucketRegion`?

The problem is since we'll be switching from the region-agnostic endpoint `s3.amazonaws.com` to a region-specific endpoint `s3.${region}.amazonaws.com` (for non-Chinese regions`, this would be a breaking change for users that have a bucket in another region unless we know what region that bucket resides in. And the only way of doing that consistently is via `HEAD` / curl. If you add a new option, then users would need to set it before they can continue using this plugin. What's wrong with the `curl` / HTTP HEAD approach?

Reply:
Sorry, I meant split stacks. Too many plugins :) (and typing on phone while getting kids to school)

It would be a custom option we allow them to provide, if they provide it, we look up the region endpoint using the aws-info module

```yml
custom:
  splitStacks:
     bucketRegion: 'x-yz-a'
```

So existing users, not specifying it, use the agnostic endpoint.

It's more of a manual step, but doesn't break if AWS changes the behaviour of the HEAD operation.

I'm fine with either, and I'm not too picky on http libraries (I've used `got` before? but that's not an endorsement per se)

Reply:
I see your point. I checked the [botocore implementation](https://github.com/boto/botocore/blob/develop/botocore/utils.py#L1178) for handling redirects and they do the same thing to get a bucket region. This code was added [3 years ago](https://github.com/boto/botocore/commit/28b8831f5f8518bef5accf994fc377c58a3569f7) and not touched since so I suggest we are safe to implement it? :) It is the most robust.

The drawback with adding the config is that it is one more step for the user and imho not very intuitive. The HEAD approach will just work for every region. Which is nice.

Reply:
Sounds good!  Sorry for the scope creep but do you mind generating URLs in the v2 format since AWS is deprecated the old: https://forums.aws.amazon.com/ann.jspa?annID=6776

>  (example: //<bucketname>.s3.amazonaws.com/key

Reply:
That's interesting. First of all, thanks for bringing this up, I knew those 2 ways of achieving this existed, but I didn't know that one was about to be deprecated.

Unfortunately, if the bucket name is not DNS compliant (i.e. contains a dot), this request will run into an SSL error. That was my original reason. Apparently that's why it's also [recommended against](https://www.cloudconformity.com/conformity-rules/S3/dns-compliant-buckets-names.html) to incorporate them in bucket names.

My question is could this be a breaking change? So the deployment bucket that serverless creates will never contain a dot, but you can specify an arbitrary deployment bucket in your `serverless.yml` configuration that could contain a period. My suggestion would be to do something similar to the [AWS CLI](https://github.com/boto/botocore/blob/develop/botocore/utils.py#L855) in the command `aws s3api head-bucket --debug --bucket non.compliant`. The debug output contains this:

```
2019-05-03 22:32:11,880 - MainThread - botocore.hooks - DEBUG - Event before-sign.s3.HeadBucket: calling handler <function fix_s3_host at 0x11009c840>
2019-05-03 22:32:11,880 - MainThread - botocore.utils - DEBUG - Checking for DNS compatible bucket for: https://s3.eu-central-1.amazonaws.com/non.compliant
2019-05-03 22:32:11,880 - MainThread - botocore.utils - DEBUG - Not changing URI, bucket is not DNS compatible: non.compliant
```

So I'd emit a loud warning if we encounter a bucket name with a dot and use the v1 URI (and maybe set the `Host` header?). If the bucket name doesn't contain a dot, we can simply use the v2 URI. Thoughts?

Reply:
Yep sounds good (I replied via email but that doesn't go to the thread it seems)